File: LeetCode/409. Longest Palindrome/409.java
class Solution {
    // Define a class named Solution

    public int longestPalindrome(String s) {
        // Define a method named longestPalindrome that takes a string s as input

        int[] freq = new int[128];
        // Initialize an array freq of size 128 to store the frequency of each ASCII character in s

        for (char c : s.toCharArray()) {
            freq[c]++;
        }
        // Iterate over each character c in s, incrementing its count in the freq array

        int ans = 0;
        // Initialize a variable ans to 0. This variable will store the length of the longest palindrome that can be built with the characters of s

        for (int f : freq) {
            ans += f / 2 * 2;
        }
        // For each value f in the freq array, add the largest even number less than or equal to f to ans

        return ans < s.length() ? ans + 1 : ans;
        // If ans is less than the length of s, return ans + 1. This accounts for the fact that a palindrome can have one character that appears an odd number of times in the middle. Otherwise, return ans.
    }
}

/*
 class Solution {
    public int longestPalindrome(String s) {
        int[] arr= new int[100];
        for(int i=0;i<s.length();i++){
            arr[s.charAt(i)-'A']++;
        }
        int ans=0;
        for(int i=0;i<100;i++){
            if(arr[i]%2==1){
                ans+=(arr[i]-1);
            }else{
                ans+=arr[i];
            }
        }
        if(ans<s.length()){
            return ans+1;
        }
        return ans;
    }
}
 */

 /*
  class Solution {
    public int longestPalindrome(String s) {
        int [] count = new int[128];
        for(char c:s.toCharArray()){
            count[c]++;
        }
        int ans=0;
        for(int f:count){
            ans+= f%2==0 ? f : f-1;
        }

        boolean odd = Arrays.stream(count).anyMatch(f -> f%2 ==1);
        return ans+(odd ? 1 : 0);
    }
}
  */
  /*
   class Solution {
    public int longestPalindrome(String s) {
        Map<Character, Integer> lower = new HashMap<>();
        int sum = 0;
        boolean flag = false;

        for (char c : s.toCharArray()) {
            lower.put(c, lower.getOrDefault(c, 0) + 1);
        }

        for (Character c : lower.keySet()){
            if (lower.get(c) % 2 == 0){
                sum += lower.get(c);
            } else {
                sum += lower.get(c) - 1;
                flag = true;
            }
        }

    


        return sum + (flag ? 1 : 0);







    }
}
   */
  /*
   class Solution {
    public int longestPalindrome(String s) {
        int oddCount = 0;
        Map<Character, Integer> map = new HashMap<>();
        for (char ch : s.toCharArray()) {
            map.put(ch, map.getOrDefault(ch, 0) + 1);
            if (map.get(ch) % 2 == 1)
                oddCount++;
            else
                oddCount--;
        }
        if (oddCount > 1)
            return s.length() - oddCount + 1;
        return s.length();
    }
}
   */

File: LeetCode/409. Longest Palindrome/409.py
class Solution(object):
    # Define a class named Solution

    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: int
        """
        # Define a method named longestPalindrome that takes a string s as input

        count = {}
        # Initialize an empty dictionary count to store the frequency of each character in s

        for c in s:
            count[c] = count.get(c, 0) + 1
        # Iterate over each character c in s, incrementing its count in the count dictionary

        length = 0
        # Initialize a variable length to 0. This variable will store the length of the longest palindrome that can be built with the characters of s

        for v in count.values():
            length += v // 2 * 2
            # For each value v in the count dictionary, add the largest even number less than or equal to v to length

            if length % 2 == 0 and v % 2 == 1:
                length += 1
        # If length is even and v is odd, increment length by 1. This accounts for the fact that a palindrome can have one character that appears an odd number of times in the middle

        return length
        # Return the length of the longest palindrome that can be built with the characters of s

#other people's solution
'''
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: int
        """
        lookup = Counter([char for char in s])
        pal_len= 0
        odd = False
        for char,value in lookup.items():
            if not odd and value%2 != 0:
                pal_len +=value
                print("odd{}",format(value))
                odd = True
            else:
                pal_len += int(value/2)*2
        return pal_len  
        
'''
'''
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: int
        """
       
        # dic = {}
        # for char in s:
        #     dic[char] = dic.get(char, 0) + 1
        
        # counts = dic.values()
    
        # odd_flag = 0
        # res = 0
        # for x in counts:
        #     if x % 2 == 0:
        #         res += x
        #     else:
        #         res += x - 1
        #         odd_flag = 1
        # return res + odd_flag

        ss = set()
        for char in s:
            if char not in ss:
                ss.add(char)
            else:
                ss.remove(char)
        
        if len(ss) > 0:
            return len(s) - len(ss) + 1
        
        return len(s)
'''
'''
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: int
        """
        if len(s)==1:
            return 1
        
        char_count = {}

        for letter in s:
            if letter in char_count:
                char_count[letter]+=1
            else:
                char_count[letter]=1
        odd_sum = 0
        has_odd = False
        sum_even = 0
        for num in char_count.values():
            if num%2==1:
                odd_sum+=num-1
                has_odd=True
            else:
                sum_even+=num
        if has_odd:
            return odd_sum+sum_even+1
        return sum_even
'''
'''
class Solution(object):
    def longestPalindrome(self, s):
        setA = set()
        for char in s:
            if char not in setA:
                setA.update(char)
            else:
                setA.remove(char)
        
        if len(setA) != 0:
            return len(s) - len(setA) + 1
        else:
            return len(s)
'''